Branch-SDK/src/main/java/io/branch/referral/util/BranchEvent.java

```diff
@@ -256,6 +269,175 @@ public boolean logEvent(Context context) {
         return isReqQueued;
     }
 
+    /**
+     * Logs a Branch Commerce Event based on an in-app purchase
+     *
+     * @param context  Current context
+     * @param purchase Respective purchase
+     */
+    public void logEventWithPurchase(Context context, @NonNull Purchase purchase) {
+        List<String> productIds = purchase.getProducts();
+        List<QueryProductDetailsParams.Product> productList = new ArrayList<>();
+        List<QueryProductDetailsParams.Product> subsList = new ArrayList<>();
+
+        for (String productId : productIds) {
+            QueryProductDetailsParams.Product inAppProduct = QueryProductDetailsParams.Product.newBuilder()
+                    .setProductId(productId)
+                    .setProductType(BillingClient.ProductType.INAPP)
+                    .build();
+
+            QueryProductDetailsParams.Product subsProduct = QueryProductDetailsParams.Product.newBuilder()
+                    .setProductId(productId)
+                    .setProductType(BillingClient.ProductType.SUBS)
+                    .build();
+
+            productList.add(inAppProduct);
+            subsList.add(subsProduct);
+        }
+
+        QueryProductDetailsParams queryProductDetailsParams =
+                QueryProductDetailsParams.newBuilder()
+                        .setProductList(productList)
+                        .build();
+
+        QueryProductDetailsParams querySubsProductDetailsParams =
+                QueryProductDetailsParams.newBuilder()
+                        .setProductList(subsList)
+                        .build();
+
+        PurchasesUpdatedListener purchasesUpdatedListener = new PurchasesUpdatedListener() {
+            @Override
+            public void onPurchasesUpdated(BillingResult billingResults, List<Purchase> purchases) {
+
+            }
+        };
+
+        BillingClient billingClient = BillingClient.newBuilder(context)
+                .setListener(purchasesUpdatedListener)
+                .enablePendingPurchases()
+                .build();
+
+        billingClient.startConnection(new BillingClientStateListener() {
+            @Override
+            public void onBillingSetupFinished(BillingResult billingResult) {
```

Comment:
These two (large) blocks of code are nearly identical except for checking the product type, can you make a common function for querying product details your two lists?